### PR TITLE
clean ${PAI_WORK_DIR} before mv content to this folder

### PR DESCRIPTION
**Pull request by Binyang2014**
In prod env, we saw when node down and up. The pod running on this node
may be restarted by kubelet. Even though the `restartPolicy` set to
`Never`. If that happens, the init container script will failed when
executing `mv` command. To prevent this, we clean the `${PAI_WORK_DIR}`
before mv content to this folder.

### CODE REVIEW

**Comment by yqwang-ms** [comment]
We should make our runtime **Idempotent**, since k8s restartPolicy=Never seems to be at least once execution like FC. So please confirm it and refine other places in runtime code.

File: src/kube-runtime/src/init [hunk 1]
```diff
@@ -78,6 +79,10 @@ PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
 
 PAI_LOG_DIR=${PAI_WORK_DIR}/logs/${FC_POD_UID}
 
+# Clean ${PAI_WORK_DIR} and ${PAI_LOG_DIR} since they may contain last execution content. (rarely happen, but seen in real world)
+rm -rf ${PAI_WORK_DIR}/*
+rm -rf ${PAI_LOG_DIR}/*
```
Comment (abuccts) [inline src/kube-runtime/src/init]: do you need to clean log dir? keep the previous log may be helpful

File: src/kube-runtime/src/init [hunk 2]
```diff
@@ -78,6 +79,10 @@ PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
 
 PAI_LOG_DIR=${PAI_WORK_DIR}/logs/${FC_POD_UID}
```
Comment (yqwang-ms) [inline src/kube-runtime/src/init]: We may need to keep all history and logs for tracking. So to achieve Idempotent, can you use a different folder each time?

Comment (Binyang2014) [reply src/kube-runtime/src/init]: Will move old logs to ${PAI_LOG_DIR}/pre_{index}

---
In reply to: [332809405](https://github.com/microsoft/pai/pull/3695#discussion_r332809405) [](ancestors = 332809405)

File: src/kube-runtime/src/init (hunk 1 above)
Comment (Binyang2014) [reply src/kube-runtime/src/init]: Sure, will change the code

---
In reply to: [332808965](https://github.com/microsoft/pai/pull/3695#discussion_r332808965) [](ancestors = 332808965)

File: src/kube-runtime/src/init [hunk 3]
```diff
@@ -53,14 +60,15 @@ function exit_handler()
...
-
-  echo "Unknown exit code, platform error"
-  exit 1
+  case $EXIT_CODE in
+    130|131|132|134|135|136|137|139|141|143)
+      exit $EXIT_CODE
+      ;;
+    *)
+      echo "Unknown exit code, platform error"
+      exit 248
```
Comment (yqwang-ms) [inline src/kube-runtime/src/init]: Expose or log orignal exitcode

File: src/kube-runtime/src/init [hunk 4]
```diff
@@ -78,6 +86,17 @@ PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
...
+# Move previous logs to another folder. Notice: for init.log, some part will append to previous log file
+LOG_FILES=$(find $PAI_LOG_DIR -maxdepth 1 -type f)
+if [[ ! -z $LOG_FILES ]]; then
+  PRE_LOG_DIR=$PAI_LOG_DIR/prelog_$(date +%s)
+  mkdir $PRE_LOG_DIR
+  mv $LOG_FILES $PRE_LOG_DIR
+fi
+
+# Clean ${PAI_WORK_DIR} since it may contain last execution content. (rarely happen, but seen in real world)
+rm -rf ${PAI_WORK_DIR}/*
```
Comment (yqwang-ms) [inline src/kube-runtime/src/init]: use a different folder each time: Also for PAI_WORK_DIR

File: src/kube-runtime/src/init (hunk 3 above)
Comment (Binyang2014) [reply src/kube-runtime/src/init]: logged in line 41

Comment (Binyang2014) [reply src/kube-runtime/src/init]: logged in line 41

File: src/kube-runtime/src/init (hunk 4 above)
Comment (Binyang2014) [reply src/kube-runtime/src/init]: Hmm, this folder is created by and mounted by k8s: https://github.com/microsoft/pai/blob/2a2a6e25cca3bc2b272e2faf88d000d6919b9a7b/src/rest-server/src/models/v2/job/k8s.js#L318
The folder name is fixed during pod creation. We can't assign arbitrary name for different container.

And this folder content will no be persisted and always be the PAI runtime source code. (Also this folder will be deleted when deleting pod). So I think it's not necessary to use the different folder each time

File: src/kube-runtime/src/init [hunk 5]
```diff
@@ -78,6 +86,17 @@ PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
 
 PAI_LOG_DIR=${PAI_WORK_DIR}/logs/${FC_POD_UID}
 
+# Move previous logs to another folder. Notice: for init.log, some part will append to previous log file
+LOG_FILES=$(find $PAI_LOG_DIR -maxdepth 1 -type f)
+if [[ ! -z $LOG_FILES ]]; then
+  PRE_LOG_DIR=$PAI_LOG_DIR/prelog_$(date +%s)
```
Comment (yqwang-ms) [inline src/kube-runtime/src/init]: >$(date +%s) [](start = 34, length = 11)

time in second may be the same. So why not just use the LOG_FILES count as suffix?

File: src/kube-runtime/src/init [hunk 6]
```diff
@@ -78,6 +86,22 @@ PAI_RUNTIME_DIR=${PAI_WORK_DIR}/runtime.d
 
 PAI_LOG_DIR=${PAI_WORK_DIR}/logs/${FC_POD_UID}
 
+# Move previous logs to another folder. Notice: for init.log, some part will append to previous log file
+LOG_FILES=$(find $PAI_LOG_DIR -maxdepth 1 -type f)
+if [[ ! -z $LOG_FILES ]]; then
+  RANDOM=$$
```
Comment (yqwang-ms) [inline src/kube-runtime/src/init]: Can we print warning log here?

**Pull request merged**